File: packages/typra-emitter/fixtures/features/typed-seam-conformance/vector-adapters/swift/TypedConformanceTests.swift
// Committed consumer double for the Swift typed @vector conformance entrypoint
// (issue #511 Cat 1, typra#306 Track A). This is the WHOLE authored surface a
// consumer needs to migrate a plain seam off the stringly VectorAdapters
// registry: a real `Transformer` protocol impl and one typed call to the
// emitted `runTransformerConformance`. No registry, no string keys, no per-op
// marshalling double.
//
// `runTransformerConformance` is emitted only when the entrypoint lands; on
// `main` the symbol does not exist, so this file fails to compile (the red-first
// signal). Conforming `TransformerImpl` to the emitted `Transformer` protocol
// obliges every op, so a forgotten method fails to COMPILE — completeness is not
// a runtime lookup.
import XCTest

@testable import TypraFixturesFeaturesTypedSeamConformance

private enum TransformerError: Error, CustomStringConvertible {
  case boom
  var description: String { "boom not allowed" }
}

private struct TransformerImpl: Transformer {
  func transform(text: String) async throws -> String {
    if text == "boom" { throw TransformerError.boom }
    return text.trimmingCharacters(in: .whitespaces)
  }
}

// A model-in/model-out seam impl: proves the typed entrypoint decodes the Note
// param via the emitted `Note.load(...)` and serializes the Note result back
// through `try actual.save()`. Uppercases the title and returns the value.
private struct ReviserImpl: Reviser {
  func revise(note: Note) async throws -> Note {
    var revised = note
    revised.title = note.title.uppercased()
    return revised
  }
}

// An array-of-model-in/array-of-model-out seam impl: proves the typed entrypoint
// decodes each `[Note]` element via `Note.load(...)` and serializes the `[Note]`
// result back through `try $0.save()`. Reverses the array.
private struct CollatorImpl: Collator {
  func collate(notes: [Note]) async throws -> [Note] {
    return notes.reversed()
  }
}

// A carrier-param seam impl: the note flows through typed while `options` is an
// untyped `[String: Any]` carrier (prompty's `Renderer.render` `inputs`).
// `reassemble` takes the OPTIONAL carrier (`Parser.parse` `context?`), which
// lowers to `[String: Any]?`; the absent-carrier vector proves an omitted
// optional carrier decodes to `nil`. The impls ignore `options` — the point is
// only that the untyped param decodes and the `[Note]` return still compares.
private struct AssemblerImpl: Assembler {
  func assemble(note: Note, options: [String: Any]) async throws -> [Note] {
    return [note]
  }

  func reassemble(note: Note, options: [String: Any]?) async throws -> [Note] {
    return [note]
  }
}

final class TypedConformanceTests: XCTestCase {
  func testTransformerTypedConformance() async throws {
    try await runTransformerConformance(TransformerImpl())
  }

  func testReviserTypedConformance() async throws {
    try await runReviserConformance(ReviserImpl())
  }

  func testCollatorTypedConformance() async throws {
    try await runCollatorConformance(CollatorImpl())
  }

  func testAssemblerTypedConformance() async throws {
    try await runAssemblerConformance(AssemblerImpl())
  }
}

File: packages/typra-emitter/fixtures/features/typed-seam-conformance/vector-adapters/csharp/TypedConformanceTests.cs
// Committed consumer double for the C# typed @vector conformance entrypoint
// (issue #511 Cat 1). This file is NOT emitted — it is what a real consumer
// authors: genuine typed seam impls plus a single typed call into the emitted
// VectorConformance.Run<Seam>ConformanceAsync. Because the entrypoint takes the
// I<Seam> interface, a forgotten op cannot compile, so this xUnit test only
// passes once the emitter emits the entrypoint (red-first on `main`).
// Transformer covers the scalar rail; Reviser covers the Phase 2
// model-in/model-out rail (its boundary model Note is in the @serializable
// closure, so the entrypoint decodes via Note.FromJson and compares via ToJson).
#nullable enable
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Typra.Fixtures.Features.TypedSeamConformance;
using Xunit;

namespace Typra.Fixtures.Features.TypedSeamConformance.ConformanceTests;

/// <summary>A real typed Transformer impl — trims, and rejects "boom".</summary>
file sealed class TransformerImpl : ITransformer
{
    public Task<string> TransformAsync(string text)
    {
        if (text == "boom")
        {
            throw new InvalidOperationException("boom not allowed");
        }

        return Task.FromResult(text.Trim());
    }
}

/// <summary>A real typed Reviser impl — upper-cases the note title, passes the
/// body through. The note arrives decoded (Note.FromJson); mutate and return it
/// so the result stays a Note for the entrypoint's ToJson compare.</summary>
file sealed class ReviserImpl : IReviser
{
    public Task<Note> ReviseAsync(Note note)
    {
        note.Title = note.Title.ToUpperInvariant();
        return Task.FromResult(note);
    }
}

/// <summary>A real typed Collator impl — reverses the notes, each passing
/// through unchanged. The notes arrive decoded (per-element Note.FromJson);
/// return a List&lt;Note&gt; so the entrypoint's per-element ToJson compare stays
/// canonical.</summary>
file sealed class CollatorImpl : ICollator
{
    public Task<List<Note>> CollateAsync(List<Note> notes)
    {
        return Task.FromResult(notes.AsEnumerable().Reverse().ToList());
    }
}

public class TypedConformanceTests
{
    [Fact]
    public async Task TransformerTypedConformance()
    {
        await VectorConformance.RunTransformerConformanceAsync(new TransformerImpl());
    }

    [Fact]
    public async Task ReviserTypedConformance()
    {
        await VectorConformance.RunReviserConformanceAsync(new ReviserImpl());
    }

    [Fact]
    public async Task CollatorTypedConformance()
    {
        await VectorConformance.RunCollatorConformanceAsync(new CollatorImpl());
    }
}
